lib/install.js

```diff
@@ -79,277 +84,214 @@ function install (fs, gyp, argv, callback) {
         return
       }
       log.verbose('install', 'version is already installed, need to check "installVersion"')
-      var installVersionFile = path.resolve(devDir, 'installVersion')
-      fs.readFile(installVersionFile, 'ascii', function (err, ver) {
+      const installVersionFile = path.resolve(devDir, 'installVersion')
+      fs.readFile(installVersionFile, 'ascii', (err, ver) => {
         if (err && err.code !== 'ENOENT') {
           return cb(err)
         }
-        var installVersion = parseInt(ver, 10) || 0
+        const installVersion = parseInt(ver, 10) || 0
         log.verbose('got "installVersion"', installVersion)
         log.verbose('needs "installVersion"', gyp.package.installVersion)
         if (installVersion < gyp.package.installVersion) {
           log.verbose('install', 'version is no good; reinstalling')
-          go()
+          go().then(cb, cb)
         } else {
           log.verbose('install', 'version is good')
           cb()
         }
       })
     })
   } else {
-    go()
+    go().then(cb, cb)
   }
 
-  function getContentSha (res, callback) {
-    var shasum = crypto.createHash('sha256')
-    res.on('data', function (chunk) {
-      shasum.update(chunk)
-    }).on('end', function () {
-      callback(null, shasum.digest('hex'))
-    })
+  class ShaSum extends stream.Transform {
+    constructor (callback) {
+      super()
+      this._callback = callback
+      this._digester = crypto.createHash('sha256')
+    }
+
+    _transform (chunk, _, callback) {
+      this._digester.update(chunk)
+      callback(null, chunk)
+    }
+
+    _flush (callback) {
+      this._callback(null, this._digester.digest('hex'))
+      callback()
+    }
   }
 
-  function go () {
+  async function go () {
     log.verbose('ensuring nodedir is created', devDir)
 
     // first create the dir for the node dev files
-    fs.mkdir(devDir, { recursive: true }, function (err, created) {
-      if (err) {
-        if (err.code === 'EACCES') {
-          eaccesFallback(err)
-        } else {
-          cb(err)
-        }
-        return
-      }
+    try {
+      const created = await fs.promises.mkdir(devDir, { recursive: true })
 
       if (created) {
         log.verbose('created nodedir', created)
       }
-
-      // now download the node tarball
-      var tarPath = gyp.opts.tarball
-      var badDownload = false
-      var extractCount = 0
-      var contentShasums = {}
-      var expectShasums = {}
-
-      // checks if a file to be extracted from the tarball is valid.
-      // only .h header files and the gyp files get extracted
-      function isValid (path) {
-        var isValid = valid(path)
-        if (isValid) {
-          log.verbose('extracted file from tarball', path)
-          extractCount++
-        } else {
-          // invalid
-          log.silly('ignoring from tarball', path)
-        }
-        return isValid
+    } catch (err) {
+      if (err.code === 'EACCES') {
+        eaccesFallback(err)
```

Comment:
This is going to result in a double call to `cb()` because you've at once handed off flow control to the returning promise while also attempting to regain control here. `eaccesFallback` calls `cb` but you also have `then(cb, cb)` so 🤷. I'm not really sure how best to resovle this without pulling it apart even further, the middle-place with half Promises and half callbacks is very unpleasant. (See also https://github.com/nodejs/node-gyp/pull/2140 which perhaps you'd be interested in contributing to if you're interested in this challenge).